panel/widgets/chatbox.py

```diff
@@ -219,9 +302,23 @@ class ChatBox(CompositeWidget):
         e.g. `[{'You': ['Welcome!', 'Good bye!']}]` The message(s) can be
         any Python object that can be rendered by Panel.""")
 
-    primary_name = param.String(default=None, doc="""Name of the primary user (the one who inputs messages);
+    primary_name = param.String(default=None, doc="""
+        Name of the primary user (the one who inputs messages);
         the first key found in value will be used if unspecified.""")
 
+    response_callback = param.Callable(default=None, doc="""
```

Comment:
Please consider that right now everyone is in a flow of an "assistant" responding to a "user". Multiple users that should be responded to. Or specialized bots responding to another specialized bots.

I.e. I believe its important to be able to be able to support multiple callbacks. One per role/ name.

That is what I implemented in https://github.com/holoviz/panel/pull/5279.

Reply:
```
import re
from typing import Any

from panel.io.mime_render import exec_with_return
import pandas as pd
import panel as pn
import openai

pn.extension()

DATAFRAME_PROMPT = """
    Here are the columns in your DataFrame: {columns}.
    Create  a plot with hvplot that highlights an interesting
    relationship between the columns with hvplot groupby kwarg.
"""

CODE_REGEX = re.compile(r"```python(.*?)```", re.DOTALL)


async def respond_with_openai(contents: Any):
    # extract the DataFrame
    if isinstance(contents, pd.DataFrame):
        global df
        df = contents
        columns = contents.columns
        message = DATAFRAME_PROMPT.format(columns=columns)

    # ask OpenAI to plot
    response = await openai.ChatCompletion.acreate(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": message}],
        temperature=0,
        max_tokens=500,
        stream=True,
    )
    message = ""
    async for chunk in response:
        message += chunk["choices"][0]["delta"].get("content", "")
        yield {"OpenAI": message}


async def respond_with_executor(code: str):
    return {"Executor": exec_with_return(code=code, global_context=globals())}


async def response_callback(
    contents: Any,
    name: str,
    chat_box: pn.widgets.ChatBox,
):
    if name == "You":
        async for chunk in respond_with_openai(contents):
            yield chunk
    elif CODE_REGEX.search(contents):
        yield await respond_with_executor(CODE_REGEX.search(contents).group(1))


pn.widgets.ChatBox(
    message_input_widgets=[pn.widgets.FileInput()], response_callback=response_callback
).servable()
```

This semi-works, but the executor responds too fast before the OpenAI response is completely finished.

<img width="1302" alt="image" src="https://github.com/holoviz/panel/assets/15331990/afdaa456-6633-4be5-891a-d3f26d1a1fd8">

Reply:
Okay, I got the multi responses working:

https://github.com/holoviz/panel/assets/15331990/e5269089-093c-42b0-b74d-cf23ed2bfc30